src/coreclr/jit/emitarm64.cpp

```diff
@@ -13510,6 +13510,20 @@ void emitter::emitInsLoadStoreOp(instruction ins, emitAttr attr, regNumber dataR
     }
     else // addr is not contained, so we evaluate it into a register
     {
+#ifdef DEBUG
```

Comment:
Nit: maybe `void Lowering::CheckNode(Compiler* compiler, GenTree* node)` is a better place for such check:
1. it is earlier;
2. it is one place for all platforms

Reply:
I'm worried about moving the assert to `CheckNode`:

- It's pretty far upstream. Am I guaranteed no other phases are going to introduce the bad pattern after that is run and before this?
- I'd have to check for any IND-like thing that might end up getting to this code path, then looking at the IND's address. Might I miss something doing that? It feels perhaps like doing too much anticipating what downstream phases and codegen are actually going to do. E.g., it looks like for STOREIND(LCL_VAR_ADDR) of a SIMD12 type, the addr isn't contained. I'd have to check that.

Comments?

Reply:
> It's pretty far upstream. Am I guaranteed no other phases are going to introduce the bad pattern after that is run and before this?

We are adding "tracked" before Lowering and "contained" during Lowering, so after Lowering both should be stable  (there is one exception for of `ClearContained` for bitcast but it is not important here)

> I'd have to check for any IND-like thing that might end up getting to this code path, then looking at the IND's address. Might I miss something doing that? It feels perhaps like doing too much anticipating what downstream phases and codegen are actually going to do. E.g., it looks like for STOREIND(LCL_VAR_ADDR) of a SIMD12 type, the addr isn't contained. I'd have to check that.

I was thinking about adding something like:
```
case GT_LCL_VAR_ADDR:
case GT_LCL_FLD_ADDR:
{
    const GenTreeLclVarCommon* varNode = tree->AsLclVarCommon();
    const LclVarDsc* varDsc  = emitComp->lvaGetDesc(varNode);
    assert(varNode->isContained() || !varDsc->lvTracked);
    break;
}
```

to `Lowering::CheckNode`.

would not it be enough?

Reply:
That doesn't work. There are cases where the LCL_VAR_ADDR is a not required to be contained. I see such a case with `STORE_OBJ(LCL_VAR_ADDR, LCL_VAR)`. I also see `STOREIND-simd12(LCL_FLD_ADDR, LCLVAR-simd16)`.

If I add `|| !varTypeIsGC(varDsc)` that's still not enough, as I see cases where we have not-contained tracked address loads, which is fine because `genCodeForLclAddr` does the right liveness updates for uses. The issue is for stores. So, I'd need to look at the context (parent node). And it's not clear that just STOREIND is necessary to check.

Bottom-line: does this change really need this more general assert in a common location? Or can I leave it as-is?